File: C/15.命令模式/src/SellStock.c
/**
 * @file SellStock.c
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2020-08-09
 *
 * @copyright Copyright (c) 2020
 *
 */

#include "OrderStruct.h"
#include "Stock.h"
#include "ctools.h"

typedef struct {
    Stock* abc_stock;
} priv_t;

static void init(struct Order* order, Stock* stock) {
    if (NULL == order) return;

    priv_t* priv = (priv_t*)calloc(1, sizeof(priv_t));
    if (NULL == priv) return;

    priv->abc_stock = stock;
    order->priv = priv;
}

static void destroy(struct Order* order) {
    if (NULL == order || NULL == order->priv) return;
    freep((void**)&order->priv);
}

static void execute(struct Order* order) {
    if (NULL == order || NULL == order->priv) return;
    priv_t* priv = (priv_t*)order->priv;
    stock_sell(priv->abc_stock);
}

struct Order* sell_stock_order_create(Stock* stock) {
    struct Order* order = (struct Order*)calloc(1, sizeof(struct Order));
    if (NULL == order) return NULL;

    init(order, stock);
    order->execute = execute;
    order->destroy = destroy;
    return order;
}

File: Objective-C/15.CommandPattern/SellStock.m
//
//  SellStock.m
//  15.CommandPattern
//
//  Created by Layne on 2020/11/5.
//

#import "SellStock.h"

@implementation SellStock {
    Stock* _abcStock;
};

- (instancetype)initWithStock:(Stock*)abcStock {
    self = [super init];
    if (self) {
        _abcStock = abcStock;
    }
    return self;
}

- (void)execute {
    [_abcStock sell];
}

@end
